File: server/entity/src/media.rs
use rocket::serde::{Deserialize, Serialize};
use sea_orm::entity::prelude::*;

use crate::util::FileStatus;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[serde(crate = "rocket::serde", rename_all = "camelCase")]
#[sea_orm(table_name = "media")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    /// The id of the series this media belongs to.
    pub series_id: i32,
    /// name of the media. ex: "The Amazing Spider-Man (2018) #69.cbz"
    pub name: String,
    /// description of the media. ex: "Spidey and his superspy sister, Teresa Parker, dig to uncover THE CHAMELEON CONSPIRACY."
    pub description: Option<String>,
    /// the size of the media in bytes.
    pub size: i64,
    /// the file extension of the media. ex: "cbz"
    pub extension: String,
    /// the number of pages in the media. ex: "69"
    pub pages: i32,
    /// the date in which the media was last updated in the FS. ex: "2020-01-01"
    #[sea_orm(column_type = "DateTime")]
    pub updated_at: Option<chrono::NaiveDateTime>,
    /// whether or not the media is downloaded to the client. ex: true
    pub downloaded: bool,
    /// the checksum hash of the file contents. Used to ensure only one instance of a file in the database
    #[sea_orm(unique)]
    pub checksum: String,
    /// the path of the media. ex: "/home/user/media/comics/The Amazing Spider-Man (2018) #69.cbz"
    pub path: String,
    /// The status of the media since last scan or access
    #[sea_orm(default_value = FileStatus::Ready)]
    pub status: FileStatus,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::series::Entity",
        from = "Column::SeriesId"
        to="super::series::Column::Id"
    )]
    Series,

    #[sea_orm(has_many = "super::read_progress::Entity")]
    ReadProgress,
}

impl Related<super::series::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Series.def()
    }
}

impl Related<super::read_progress::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::ReadProgress.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}

File: core/bindings/Media.ts
import { FileStatus } from '.';
import { Tag } from './Tag';

export interface Media {
	/**
	 * The id of the media.
	 */
	id: string;
	/**
	 * The id of the series which this media belongs to.
	 */
	seriesId: string;
	/**
	 * The common name of the media. Either the name extracted from metadata or the file name.
	 */
	name: string;
	/**
	 * The description extracted from metadata.
	 */
	description?: string;
	/**
	 * The size of the media in bytes.
	 */
	size: number;
	/**
	 * The extension of the media file.
	 */
	extension: string;
	/**
	 * The number of (image) pages in the media.
	 */
	pages: number;
	/**
	 * The date/time the media was last modified. Extracted from metadata.
	 */
	updatedAt?: Date;
	/**
	 * The checksum of the media file.
	 */
	checksum: string;
	/**
	 * The path of the media file on disk.
	 */
	path: string;
	/**
	 * The page the viewing user is currently on for the associated media.
	 */
	currentPage?: number;
	/**
	 * The status of the media.
	 */
	status: FileStatus;
	/**
	 * The user assigned tags for this media. This will be undefined only if the relation is not loaded.
	 * @see Tag
	 */
	tags?: Tag[];
}
